proxy/proxy/proxy.go

```diff
@@ -165,41 +184,68 @@ func (p *Proxy) serveConn(c net.Conn) error {
 		p.reqPool.Release(req)
 		p.BufioPool.ReleaseReader(reader)
 	}
+	defer releaseReqAndReader()
 
-	if err := req.ReadFrom(reader); err != nil {
-		releaseReqAndReader()
-		return util.ErrWrapper(err, "fail to read http request header")
-	}
-	if len(req.HostWithPort()) == 0 {
-		if e := p.writeFastError(c, http.StatusBadRequest,
-			"This is a proxy server. Does not respond to non-proxy requests.\n"); e != nil {
-			return util.ErrWrapper(e, "fail to response non-proxy request")
+	var (
+		connTime, currentTime time.Time
+		lastReadDeadlineTime  time.Time
+		connectionClose       bool
+	)
+	currentTime = servertime.CoarseTimeNow()
+	connTime = currentTime
+
+	for {
+		if p.MaxKeepaliveDuration > 0 {
+			lastReadDeadlineTime = p.updateReadDeadline(c, currentTime, connTime, lastReadDeadlineTime)
+			if lastReadDeadlineTime.IsZero() {
+				return util.ErrWrapper(nil, "exceeded MaxKeepaliveDuration, close connection after %s", p.MaxKeepaliveDuration)
+			}
 		}
-		releaseReqAndReader()
-		return nil
-	}
 
-	//handle http requests
-	if !http.IsMethodConnect(req.Method()) {
-		err := p.Handler.handleHTTPConns(c, req,
-			p.BufioPool, &p.Client)
-		releaseReqAndReader()
-		if err != nil {
-			return util.ErrWrapper(err, "error HTTP traffic %s ", req.HostWithPort())
+		if err := req.ReadFrom(reader); err != nil {
+			return util.ErrWrapper(err, "fail to read http request header")
+		}
+
+		if len(req.HostWithPort()) == 0 {
+			if e := p.writeFastError(c, http.StatusBadRequest,
+				"This is a proxy server. Does not respond to non-proxy requests.\n"); e != nil {
+				return util.ErrWrapper(e, "fail to response non-proxy request")
+			}
+			return nil
+		}
+
+		if req.ConnectionClose() && !req.ProxyConnectionKeepalive() {
```

Comment:
req's full header haven't been initialized till now:

req.ConnectionClose() returns false always 

req.ProxyConnectionKeepalive() returns false always  too

so  req.ConnectionClose() && !req.ProxyConnectionKeepalive() is always  fasle.